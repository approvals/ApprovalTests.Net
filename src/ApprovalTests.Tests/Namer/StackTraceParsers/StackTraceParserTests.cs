using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using ApprovalTests.Namers.StackTraceParsers;
using ApprovalUtilities.SimpleLogger;
using NUnit.Framework;

namespace ApprovalTests.Tests.Namer.StackTraceParsers
{
    [TestFixture]
    public class StackTraceParserTests
    {
        [Test]
        public void NamespaceFilters()
        {
            var namespaces = new[] { GetType().Namespace, "", null, typeof(string).Namespace, typeof(Approvals).Namespace, typeof(Logger).Namespace};
            Approvals.VerifyAll("Namespace is part of Approvals", namespaces, n => $"{n} = {AttributeStackTraceParser.IsNamespaceApprovals(n)}");
        }



        [Test]
        public void Parse_UsingStaticInitialize_DontThrowInvalidOperationException()
        {
            var parser = new StackTraceParser();

            try
            {
                Parallel.ForEach(Enumerable.Range(1, 20), _ =>
                {
                    try
                    {
                        var stackTrace = new StackTrace();
                        parser.Parse(stackTrace);
                    }
                    catch (InvalidOperationException e)
                    {
                        Assert.Fail(
                            "InvalidOperationException when trying to parse stacktrace. " +
                            "This is caused by the parser collection not being thread-safe. " +
                            "Original exception message : {0} and stacktrace : {1}",
                            e.Message,
                            e.StackTrace
                        );
                    }
                    // Because the current stacktrace passed to the parse method doesn't contains any trace of a compliant stacktrace parser
                    // it's normal that we receive an exception here so let's ignore it.
                    catch (Exception e)
                    {
                        if (
                            !e.Message.Contains("Approvals is not set up to use your test framework"))
                        {
                            throw;
                        }
                    }
                });
            }
            catch (AggregateException e)
            {
                // Throw the first inner exception of the AggregateException, this way NUnit shows a much clearer result.
                throw e.InnerException;
            }
        }
    }
}