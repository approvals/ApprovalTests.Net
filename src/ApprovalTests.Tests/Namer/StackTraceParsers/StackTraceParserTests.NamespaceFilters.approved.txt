Namespace is part of Approvals

ApprovalTests.Tests.Namer.StackTraceParsers = True
 = False
 = False
System = False
ApprovalTests = True
ApprovalUtilities.SimpleLogger = True
